ibis/common/tests/test_graph_benchmarks.py

```diff
@@ -1,19 +1,49 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING
+from typing import Optional
 
-from ibis.common.collections import frozendict  # noqa: TCH001
-from ibis.common.graph import Node
-from ibis.common.grounds import Concrete
+from typing_extensions import Self  # noqa: TCH002
 
-if TYPE_CHECKING:
-    from typing_extensions import Self
+from ibis.common.collections import frozendict
+from ibis.common.graph import Graph, Node
+from ibis.common.grounds import Concrete
 
 
-class MyNode(Node, Concrete):
+class MyNode(Concrete, Node):
     a: int
     b: str
     c: tuple[int, ...]
     d: frozendict[str, int]
-    e: Self
-    f: tuple[Self, ...]
+    e: Optional[Self] = None
+    f: tuple[Self, ...] = ()
+
+
+def generate_node(depth):
+    # generate a nested node object with the given depth
+    if depth == 0:
+        return MyNode(10, "20", c=(30, 40), d=frozendict(e=50, f=60))
+    return MyNode(
+        1,
+        "2",
+        c=(3, 4),
+        d=frozendict(e=5, f=6),
+        e=generate_node(0),
+        f=(generate_node(depth - 1), generate_node(0)),
+    )
+
+
+def test_generate_node():
+    for depth in [0, 1, 2, 10, 100]:
```

Comment:
Not a huge fan of for loops in tests. It makes any specific failures difficult to debug IMO. Can you make this `parametrize`d?

Reply:
Alternatively, `0`, `1`, `10` seems fine.